lxd/storage/drivers/driver_powerflex_utils.go

```diff
@@ -0,0 +1,1021 @@
+	// Since PowerFlex volumes can be 31 characters long at max, use the image fingerprint.
+	if vol.volType == VolumeTypeImage {
+		parentName = fmt.Sprintf("%s_%s", parentName[0:12], vol.ConfigBlockFilesystem())
+	}
+
+	if vol.contentType == ContentTypeBlock {
+		parentName = fmt.Sprintf("%s%s", parentName, powerFlexBlockVolSuffix)
+	} else if vol.contentType == ContentTypeISO {
+		parentName = fmt.Sprintf("%s%s", parentName, powerFlexISOVolSuffix)
+	}
+
+	// Use volume's type as storage volume prefix, unless there is an override in powerFlexVolTypePrefixes.
+	volumeTypePrefix := string(vol.volType)
+	volumeTypePrefixOverride, foundOveride := powerFlexVolTypePrefixes[vol.volType]
+	if foundOveride {
+		volumeTypePrefix = volumeTypePrefixOverride
+	}
+
+	if snapName != "" {
+		// Always use the provided snapshot name if specified.
+		out = fmt.Sprintf("%s_%s@%s", volumeTypePrefix, parentName, snapName)
+	} else {
+		if isSnapshot {
+			// If volumeName is a snapshot (<vol>/<snap>) and snapName is not set,
+			// assume that it's a normal snapshot (not a zombie) and prefix it with
+			// "snapshot_".
+			out = fmt.Sprintf("%s_%s@snapshot_%s", volumeTypePrefix, parentName, snapshotName)
+		} else {
+			out = fmt.Sprintf("%s_%s", volumeTypePrefix, parentName)
+		}
+	}
+
+	return out
+}
+
+// createNVMeHost creates this NVMe host in PowerFlex.
+// The operation is idempotent and locked using lock name powerflex.host.
+func (d *powerflex) createNVMeHost() (string, error) {
+	nqn, err := d.getHostNQN()
+	if err != nil {
+		return "", err
+	}
+
+	unlock := locking.Lock(d.state.ShutdownCtx, "powerflex.host")
+	defer unlock()
+
+	var hostID string
+
+	host, err := d.powerFlex().getNVMeHostByNQN(nqn)
+	if err != nil {
+		if !api.StatusErrorCheck(err, http.StatusNotFound) {
+			return "", err
+		}
+
+		hostname, err := d.getHostname()
+		if err != nil {
+			return "", err
+		}
+
+		hostID, err = d.powerFlex().createHost(hostname, nqn)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if hostID == "" {
+		hostID = host.ID
+	}
+
+	return hostID, nil
+}
+
+// deleteNVMeHost delets this NVMe host in PowerFlex.
+// The operation is idempotent and locked using lock name powerflex.host.
+func (d *powerflex) deleteNVMeHost() error {
+	nqn, err := d.getHostNQN()
+	if err != nil {
+		return err
+	}
+
+	unlock := locking.Lock(d.state.ShutdownCtx, "powerflex.host")
+	defer unlock()
+
+	host, err := d.powerFlex().getNVMeHostByNQN(nqn)
+	if err != nil {
+		// Skip the deletion if the host doesn't exist anymore.
+		if api.StatusErrorCheck(err, http.StatusNotFound) {
+			return nil
+		}
+
+		return err
+	}
+
+	return d.powerFlex().deleteHost(host.ID)
+}
+
+// mapNVMeVolume maps the given volume onto this host.
+func (d *powerflex) mapNVMeVolume(name string) (revert.Hook, error) {
+	revert := revert.New()
+	defer revert.Fail()
+
+	hostID, err := d.createNVMeHost()
+	if err != nil {
+		return nil, err
+	}
+
+	volumeID, err := d.powerFlex().getVolumeID(name)
```

Comment:
just a tiny nit, but you could avoid calling d.powerFlex() multiple times by assigning it to a local client var here.

Reply:
Good suggestion, looks better this way, refactored to use `client` if there is more than one call to the PowerFlex client.